cmd/clusteringress/istio/main.go

```diff
@@ -0,0 +1,150 @@
+	"github.com/knative/pkg/configmap"
+	"github.com/knative/pkg/controller"
+	"github.com/knative/pkg/signals"
+	"github.com/knative/serving/pkg/logging"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+
+	clientset "github.com/knative/serving/pkg/client/clientset/versioned"
+	informers "github.com/knative/serving/pkg/client/informers/externalversions"
+	"github.com/knative/serving/pkg/reconciler"
+	"github.com/knative/serving/pkg/reconciler/v1alpha1/clusteringress"
+	"github.com/knative/serving/pkg/system"
+)
+
+const (
+	threadsPerController = 2
+	component            = "controller-clusteringress-istio"
+)
+
+var (
+	masterURL  = flag.String("master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
+	kubeconfig = flag.String("kubeconfig", "", "Path to a kubeconfig. Only required if out-of-cluster.")
+)
+
+func main() {
+	flag.Parse()
+
+	loggingConfigMap, err := configmap.Load("/etc/config-logging")
+	if err != nil {
+		log.Fatalf("Error loading logging configuration: %v", err)
+	}
+	loggingConfig, err := logging.NewConfigFromMap(loggingConfigMap)
+	if err != nil {
+		log.Fatalf("Error parsing logging configuration: %v", err)
+	}
+	logger, atomicLevel := logging.NewLoggerFromConfig(loggingConfig, component)
+	defer logger.Sync()
+
+	// set up signals so we handle the first shutdown signal gracefully
+	stopCh := signals.SetupSignalHandler()
+
+	cfg, err := clientcmd.BuildConfigFromFlags(*masterURL, *kubeconfig)
+	if err != nil {
+		logger.Fatalf("Error building kubeconfig: %v", err)
+	}
+
+	kubeClient, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		logger.Fatalf("Error building kubernetes clientset: %v", err)
+	}
+
+	sharedClient, err := sharedclientset.NewForConfig(cfg)
+	if err != nil {
+		logger.Fatalf("Error building shared clientset: %v", err)
+	}
+
+	servingClient, err := clientset.NewForConfig(cfg)
+	if err != nil {
+		logger.Fatalf("Error building serving clientset: %v", err)
+	}
+
+	configMapWatcher := configmap.NewInformedWatcher(kubeClient, system.Namespace)
+
+	opt := reconciler.Options{
+		KubeClientSet:    kubeClient,
+		SharedClientSet:  sharedClient,
+		ServingClientSet: servingClient,
+		ConfigMapWatcher: configMapWatcher,
+		Logger:           logger,
+		ResyncPeriod:     10 * time.Hour, // Based on controller-runtime default.
+		StopChannel:      stopCh,
+	}
+
+	servingInformerFactory := informers.NewSharedInformerFactory(servingClient, opt.ResyncPeriod)
+	sharedInformerFactory := sharedinformers.NewSharedInformerFactory(sharedClient, opt.ResyncPeriod)
+
+	clusterIngressInformer := servingInformerFactory.Networking().V1alpha1().ClusterIngresses()
+	virtualServiceInformer := sharedInformerFactory.Networking().V1alpha3().VirtualServices()
+
+	// Build all of our controllers, with the clients constructed above.
+	// Add new controllers to this array.
+	controllers := []*controller.Impl{
```

Comment:
I think we'll only have one controller here, so we can get rid of the `controllers` wrapper